### PR TITLE
US10 fuse search

### PR DESCRIPTION
**Pull request by gwittm**

### CODE REVIEW

File: components/Search/FuseSearchRecipe.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  // includeMatches: false,
+  // findAllMatches: false,
+  // minMatchCharLength: 1,
+  // location: 0,
+  threshold: 0.3,
+  // distance: 100,
+  // useExtendedSearch: false,
+  // ignoreLocation: false,
+  // ignoreFieldNorm: false,
+  // fieldNormWeight: 1,
```
Comment (NikRoe) [inline components/Search/FuseSearchRecipe.js:28]: I would remove all commented out code

File: components/Search/FuseSearchRecipe.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        {" "}
+        <svg
+          xmlns="http://www.w3.org/2000/svg"
+          height="3em"
+          viewBox="0 0 576 512"
+        >
+          {" "}
+          !--! Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com
+          License - https://fontawesome.com/license (Commercial License)
+          Copyright 2023 Fonticons, Inc. --{" "}
```
Comment (NikRoe) [inline components/Search/FuseSearchRecipe.js:73]: I would remove this

File: components/Search/FuseSearchRecipe.js
Comment (NikRoe) [inline components/Search/FuseSearchRecipe.js:1]: you are fetching your api route (`/api/recipes`) twice.

Once in your useEffect and once using useSWR.
This is not very efficient.

You should decide on having either only the useEffect and fetch or the useSWR fetch.

When only working with useEffect and fetch, you could create your own isLoading state that is true when the fetch begins and gets set to false once the fetch has been completed.

When working with useSWR you can pass an options object as a parameter. This object can have a key `onSuccess` that contains a callback function as its value. This is where you have access to the fetched data and set the fuse. 
It looks similar to this:

```js
const { data: recipes, isLoading, error } = useSWR(`/api/recipes`, {
    onSuccess: (fetchedRecipes) => {
      setFuse(new Fuse(fetchedRecipes, fuseOptions));
    },
  });
```

Please decide on either using useSWR or the useEffect with fetch and change your code according to your decision.

File: components/Search/FuseSearchRecipe.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+          <path d="M240 144A96 96 0 1 0 48 144a96 96 0 1 0 192 0zm44.4 32C269.9 240.1 212.5 288 144 288C64.5 288 0 223.5 0 144S64.5 0 144 0c68.5 0 125.9 47.9 140.4 112h71.8c8.8-9.8 21.6-16 35.8-16H496c26.5 0 48 21.5 48 48s-21.5 48-48 48H392c-14.2 0-27-6.2-35.8-16H284.4zM144 80a64 64 0 1 1 0 128 64 64 0 1 1 0-128zM400 240c13.3 0 24 10.7 24 24v8h96c13.3 0 24 10.7 24 24s-10.7 24-24 24H280c-13.3 0-24-10.7-24-24s10.7-24 24-24h96v-8c0-13.3 10.7-24 24-24zM288 464V352H512V464c0 26.5-21.5 48-48 48H336c-26.5 0-48-21.5-48-48zM48 320h80 16 32c26.5 0 48 21.5 48 48s-21.5 48-48 48H160c0 17.7-14.3 32-32 32H64c-17.7 0-32-14.3-32-32V336c0-8.8 7.2-16 16-16zm128 64c8.8 0 16-7.2 16-16s-7.2-16-16-16H160v32h16zM24 464H200c13.3 0 24 10.7 24 24s-10.7 24-24 24H24c-13.3 0-24-10.7-24-24s10.7-24 24-24z" />
+        </svg>
+        <h2>We are still busy in the kitchen! Recipes are loading...</h2>
+      </>
+    );
+  if (error) return <h2>An error occured...</h2>;
+
+  const alphabeticallySortedRecipes = recipes.sort(function (a, b) {
+    return a.title < b.title ? -1 : a.title > b.title ? 1 : 0;
+  });
```
Comment (NikRoe) [inline components/Search/FuseSearchRecipe.js:83]: chained ternary operators are really hard to read. How about this:

```suggestion
const alphabeticallySortedRecipes = recipes.toSorted((a, b) => {
  if (a.title < b.title) {
    return -1;
  }
  if (a.title > b.title) {
    return 1;
  }
  return 0;
});
```

and `.toSorted()` is a new array method that does the same as `.sort()` while not mutating the original array.

**Review by NikRoe — CHANGES_REQUESTED** [review]
Hi there,

great work and search bar 👏 

I noticed that you fetch your recipes twice and left you a comment to make this part of your code more efficient.

Please take a look at it :)

Best regards 👋

File: components/Search/FuseSearchRecipe.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        setRecipes(fetchedRecipes);
+        setIsLoading(false);
+        setFuse(new Fuse(fetchedRecipes, fuseOptions));
+      } catch (error) {
+        setError(error);
+        setIsLoading(false);
+      }
+    }
+    getRecipes();
+  }, []);
```
Comment (NikRoe) [inline components/Search/FuseSearchRecipe.js:42]: If an error occurs, you currently don't reset it anywhere.

I would recommend changing your code to this:

```suggestion
useEffect(() => {
  async function getRecipes() {
    // Reset the error state at the beginning of each call
    setError(null);
    setIsLoading(true);

    try {
      const response = await fetch(`/api/recipes`);
      const fetchedRecipes = await response.json();

      setRecipes(fetchedRecipes);
      setFuse(new Fuse(fetchedRecipes, fuseOptions));
    } catch (fetchError) {
      setError(fetchError);
    } finally {
      // Always turn off loading indicator, regardless of the outcome
      setIsLoading(false);
    }
  }

  getRecipes();
}, []);
```

**Review by NikRoe — CHANGES_REQUESTED** [review]
Hi there,

well done! I like that you chose the useEffect approach over the simple copying of the useSWR approach I suggested 👏 

I think your useEffect can be improved by resetting the error state and I left a suggestion that also gets rid of some repetition.

Ergo: You are almost there, only one teeny tiny change left 🙌

**Comment by gwittm** [comment]
Thank you for your suggestions :)

**Review by NikRoe — APPROVED** [review]
Nicely done

**Pull request merged**